sklearn/decomposition/_kernel_pca.py

```diff
@@ -210,6 +265,30 @@ def _fit_transform(self, K):
                                                 tol=self.tol,
                                                 maxiter=self.max_iter,
                                                 v0=v0)
+        elif eigen_solver == 'randomized':
+            # not using flip_sign=True for deterministic output, we do it later
+            U, S, V = randomized_svd(K, n_components=n_components,
+                                     n_iter=self.iterated_power,
+                                     flip_sign=False,
+                                     random_state=self.random_state)
+            self.alphas_ = U[:, :n_components]  # eigenvectors
+            self.lambdas_ = S[:n_components]    # eigenvalues
+
+            # Conversion of Singular values into Eigenvalues:
```

Comment:
I'm a bit confused:

- aren't the singular values the sq roots of the eigen values?
- aren't the eigen values >= 0 since i) they're the square of something or ii) the matrix K is definite positive?

Reply:
No, the singular values are not the square roots of the eigenvalues. You probably think of this because there is a link between the singular values of M and the eigs of M*M. But there is also a link between the SVD and eigs decomposition of M. See the last part of this paragraph: https://en.wikipedia.org/wiki/Singular_value_decomposition#Relation_to_eigenvalue_decomposition

Yes, in the case of PSD matrices the eigenvalues are positive. However this is kernel-based, so the gram matrix can be passed manually, and thus may actually not correcpond to a valid kernel. Since SVD runs fine for all matrices, we have to check the actual PSDefinition.

Reply:
> the gram matrix can be passed manually, and thus may actually not correcpond to a valid kernel

Yes, we should assume `K` is a proper PSD matrix.

Reply:
We even do 

```
        self.lambdas_ = _check_psd_eigenvalues(self.lambdas_,
                                               enable_warnings=False)
```

below which would properly raise a warning if K wasn't PSD

So I don't see the point in "fixing" the sign of the eigenvalues before calling this. This is even counter productive since it would make `_check_psd_eigenvalues` not raise, while it should

Reply:
I think that you are confused with singular values definition. They key point to understand is that SVD decomposition is not unique. The eigenvalue decomposition gives a particular SVD, but many others exist: simply shift signs in any two terms (on the Us, on the Vs, on the sigma) and you get another SVD.

So from a given SVD (found by randomized method), to find back the eigenvalue decomposition, we have to correct the signs so that we make sure that U = V.

Reply:
Note that I saw a lot of pages on the internet arguing that when the matrix is symmetric, then U == V. I think that **this is not true**. Consider the easy case of M being a diagonal matrix with all negative values on the diagonal. Then one SVD decomposition of M is I*(-M)*(-I) .

Reply:
Does removing this line break some tests? If yes, which ones? If no, please remove.

Long version:

K is expected to be a PSD matrix. So its eigenvalues are expected to be non negative. So there's no point in flipping their sign since they're all positive already.

K is PSD because it's of the form `A At` where `A` is the the kernel.

AAt is PSD because for any non null vector a, we have `at AAt a = (Ata)t (Ata)` = squared norm of Ata which is a non negative quantity.

Also, taking the SVD of A:

A = U S Vt
which lead to AAt = U S Vt V S Ut = U S^2 Ut.

So the lambdas_ you get are indeed the square of something and they can't be negative.

Reply:
Thanks a lot for challenging me :) 

I added the required test (`test_precomputed_kernel_not_psd`). This test fails when the lines you mention are removed. I hope this clarifies everything.

Long version: as for everywhere in scikit-learn, we have absolutely no guarantee that K is PSD when the user works in "precomputed" mode, or passes a custom kernel python function. This is the general idea behind the kernel trick: it works for all kernel methods / gram matrices but they should be valid kernels / valid gram matrices. Checking such validity is impossible upfront (users can pass any arbitrary lines of python code, or arbitrary badly designed gram matrices - most often thinking these are valid), however we **can** check this when working on the eigenvalues afterwards. The issue we face here is that, before the checking step, we use this randomized SVD solver that always returns positive singular values. So we may check the wrong thing.

Reply:
Oh by the way if you agree (from the test) that now this topic is closed. 

To prevent others to make the same mistake, I suggest to move this portion of code to a new function in `extmath` that we would name `randomized_eigs`. I'll push a proposal in a minute.

Reply:
And for the record, I compiled all the knowledge I gathered to answer you today in https://math.stackexchange.com/a/3683742/769194

Reply:
Thanks for the test. I understand your point better now.

However, as far as I understand, the proposed flip sign solution only really works if you assume `K` to be symmetric in the first place. I might be wrong, but I don't think this is going to help when `K` is an arbitrary matrix?

In any case, consider the following:
```py
import numpy as np 
from sklearn.decomposition import KernelPCA

K = [  # a non symmetric matrix
    [1, 2],
    [3, 40]
]

for solver in ["auto", "dense", "arpack", "randomized"]:
    print(solver)
    kpca = KernelPCA(kernel="precomputed", eigen_solver=solver, n_components=1)
    kpca.fit(K)
```

This code actually passes without any warning or error.

This means that even in master, we don't really ensure that `K` is a PSD matrix. The call to `_check_psd_eigenvalues` is mostly here to fix numerical issues, but by itself, it's not sufficient to make sure `K` was PSD.


I agree with you that, ideally, we should validate user input and check that `K` is PSD. However, we don't have an easy, general check for that. 


In order to move this PR forward, I would propose to remove the current sign flip, possibly with a comment e.g. "note that if K wasn't PSD, then the singular values may not be equal to the eigenvalues". The reason being that not having the sign flip check isn't worse than what we're already doing (if that makes sense).

If you deem relevant, I'm happy to discuss more in details the validation of user-provided matrices in another issue (feel free to open one). And I'm definitely open to the idea of ultimately having the sign flip check in. But we've both spent a lot of time on this PR (and again, thank you for your patience), and at this point I think it's better to leave these details out of the discussion and move forward.

> And for the record, I compiled all the knowledge I gathered to answer you today in math.stackexchange.com/a/3683742/769194

Just when you mentioned this I was literally reading that post thinking "OK that's what I need", not noticing it was you ^^

Reply:
> we've both spent a lot of time on this PR (and again, thank you for your patience), and at this point I think it's better to leave these details out of the discussion and move forward.

I understand. I just pushed something, let me know if you're ok: I moved all this debateable code in `extmath` in a `randomized_eigsh` method. That way the kernel pca code remains pure of all these considerations (`randomized_eigsh` is just an alternate to `eigsh` or `linalg.eigh`, this is easy to read/maintain).

Sorry, that makes a little bit more work to converge, but I really feel that this will make the PR closing much easier. Let me know what you think.

Reply:
oh and concerning the non symmetric kernels, I think that there is an auto-centering method (kernel_centerer) that turns them into symmetric that's why you do not see the error. but I have to check though.

Reply:
I confirm: debugging your code step by step you'll find that the first line in `_fit_transform` transforms your asymmetric kernel into a symmetric one. So it is normal that you do not get an error.

After a night of sleep over this, I came to the following conclusions:

 - for me consistency with legacy behaviour is key. So since all other methods would raise an error when the kernel is not PSD, then we have to.
 - this is especially true now that our new method is the new default selection of 'auto' mode in many cases

So I *really* insist on keeping these lines and the associated test, it is very important to prevent mistakes.

Now I propose two ways to move forward:

 - First option, pushed in the PR now: I moved the code to a new method `randomized_eigsh` in extmath, sitting just next to `randomized_svd`. This new method is similar to the other `eigsh`-like methods, except that it will propose a special `selection` parameter to choose if the n components to return should be the largest by value, or by module (absolute value). The default behaviour should be consistent with the other eigsh methods, so it should be 'value'. Since we do not need it yet though, I propose to let it raise a `NotImplementedError`. The alternate behaviour proposed, `'module'`, is the one used by kernel PCA by default.

 - second option: I do not create the new function and leave the code in kernel PCA, as it was yesterday.

Please take your time to weight the pros and cons, after all we are not in a hurry :) Thanks so much @NicolasHug for all the time spent to review. I am sure that this will constitute a great improvement for sklearn users !

Reply:
> transforms your asymmetric kernel into a symmetric one

Which proves my point that we don't do proper input validation, since no error is raised when the mapping isn't symmetric.

Sorry @smarie, there's a finite amount of time I can dedicate to this PR, and I've reached this point now.

Reply:
I think that kernel centering is actually a "feature" of scikit learn. But anyway.

> Sorry @smarie, there's a finite amount of time I can dedicate to this PR, and I've reached this point now.

I understand @NicolasHug . Thanks a lot for all the time spent already ! I'll close this thread now and conclude in the general PR conversation thread.